### ISSUE TITLE
regression: [6bd3f0865a713e641e37c98a391696a5383abd4e] Add Hurd support 

### ISSUE THREAD

**Issue by fabbione** [ClusterLabs/libqb#229]
above commit introduces a regression and corosync master is unable to start.

CPU is spinning 100% in:

18315 sched_setscheduler(18316, SCHED_RR, { 0 }) = -1 EINVAL (Invalid argument)
18315 futex(0x7fb56aebdca0, FUTEX_WAKE_PRIVATE, 1) = 1
18315 futex(0x7fb566ccb9d0, FUTEX_WAIT, 18316, NULL <unfinished ...>
18316 <... futex resumed> )             = 0
18316 --- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0} ---
18316 write(7, "\v\0\0\0", 4)           = 4
18316 rt_sigreturn()                    = 0
18316 --- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0} ---
18316 write(7, "\v\0\0\0", 4)           = 4
18316 rt_sigreturn()                    = 0
18316 --- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0} ---
18316 write(7, "\v\0\0\0", 4)           = 4
18316 rt_sigreturn()                    = 0

**Comment by fabbione** [comment]
diff --git a/lib/log_thread.c b/lib/log_thread.c
index a0cd25c..c67c294 100644
--- a/lib/log_thread.c
+++ b/lib/log_thread.c
@@ -164,11 +164,7 @@ qb_log_thread_start(void)
 
        if (logt_sched_param_queued) {
                res = qb_log_thread_priority_set(logt_sched_policy,
-#if defined(HAVE_PTHREAD_SETSCHEDPARAM) && defined(HAVE_SCHED_GET_PRIORITY_MAX)
                                                 logt_sched_param.sched_priority);
-#else
-                                                0);
-#endif
                if (res != 0) {
                        goto cleanup_pthread;
                }

this is the problematic chunk of code. HAVE_SCHED_GET_PRIORITY_MAX is never set. configure.ac doesn´t even check for it and the code does not use sched_get_priority_max anywhere anyway.

the check for HAVE_PTHREAD_SETSCHEDPARAM is redundant because qb_log_thread_priority_set is already wrapped in #if defined HAVE_......

**Comment by jnpkrn** [comment]
On 18/10/16 19:54 -0700, Fabio M. Di Nitto wrote:
> HAVE_SCHED_GET_PRIORITY_MAX is never set. configure.ac doesn´t even
> check for it and the code does not use sched_get_priority_max
> anywhere anyway.

This will likely be a readd-bad-pattern-after-it-has-been-removed 
race with
https://github.com/ClusterLabs/libqb/pull/224/commits/db7dcd141182da44a99e850699b31bcd28c68e8a

Unfortunately, it's a design flaw of Travis CI that PRs are tested
automatically only once they are added/updated, but not when the
surrounding code changes or at least somehow before the merge.

There are projects to address this timing issue with running tests
before the merge itself: http://homu.io/.  This assumes any issues
akin to what's observed with corosync could be spotted there for
it to be helpful in this case.  This is likely not the case, so
we should also isolate the reproducer from corosync into libqb's
testsuite.

-- 
Jan (Poki)

**Comment by jnpkrn** [comment]
The reported issue has its merit, not disputing that.

However it exposed another issue in corosync-libqb interaction,
that deserves further investigation:

- corosync calls qb_log_fini in corosync_exit_error, incl. case
  that scheduling priority cannot be set for logging thread

- qb_log_fini has its own problem in accessing something that it
  shouldn't, likely as a consequence of failed qb_log_thread_start
  (which in turn fails because of failed qb_log_thread_priority_set),
  which generates SIGSEGV signal

- corosync catches SIGSEGV, but in response, it calls qb_log_fini

- qb_log_fini generates another SIGSEGV as previously, boom,
  there's an infloop eating the processor as observed


So the fix should be two-edged:

- libqb should be careful about the situation qb_log_thread_priority_set
  fails from under qb_log_thread_start (perhaps by setting
  `wthread_active = QB_FALSE`?)

- corosync should prevent signal handler being called repeatedly
  (hence possibly recursively as it is likely now)

**Comment by jnpkrn** [comment]
For libqb, the issue is with:

```
==16936== Invalid read of size 4
==16936==    at 0x4C1C8C4: qb_thread_lock (util.c:69)
==16936==    by 0x4C277CB: qb_logt_worker_thread (log_thread.c:81)
==16936==    by 0x3890007554: start_thread (pthread_create.c:333)
==16936==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
```

The best solution for that seems to call `pthread_attr_setschedparam`
ahead of `pthread_create`, and only if that succeeds call `pthread_create`
with appropriately configured `pthread_attr_t` struct instead of `NULL`.


-- and/or --

`sem_post(&logt_print_finished)` should not be called from
`cleanup_pthread` goto label.  The thread should be terminated
instead?


-- and/or --

```
logt_wthread_lock = qb_thread_lock_create(QB_THREAD_LOCK_SHORT);
```

should be called regardless of `qb_log_thread_priority_set` success,
as `logt_wthread_lock` is the static pointer that may not be
initialized at the time of access from the new thread, which is
what generates SIGSEGV.

**Comment by chrissie-c** [comment]
On 19/10/16 03:54, Fabio M. Di Nitto wrote:
> diff --git a/lib/log_thread.c b/lib/log_thread.c
> index a0cd25c..c67c294 100644
> --- a/lib/log_thread.c
> +++ b/lib/log_thread.c
> @@ -164,11 +164,7 @@ qb_log_thread_start(void)
> 
> |if (logt_sched_param_queued) { res =
> qb_log_thread_priority_set(logt_sched_policy, |
> 
> -#if defined(HAVE_PTHREAD_SETSCHEDPARAM) &&
> defined(HAVE_SCHED_GET_PRIORITY_MAX)
> logt_sched_param.sched_priority);
> -#else
> 
>   * 0); -#endif if (res != 0) { goto cleanup_pthread; }
> 
> this is the problematic chunk of code. HAVE_SCHED_GET_PRIORITY_MAX is
> never set. configure.ac doesn´t even check for it and the code does not
> use sched_get_priority_max anywhere anyway.
> 
> the check for HAVE_PTHREAD_SETSCHEDPARAM is redundant because
> qb_log_thread_priority_set is already wrapped in #if defined HAVE_......
> 


I've reverted that bit of the commit and corosync behaves now.

Poki is right in that it would be nice if corosync/libqb didn't get into
this segv loop - it's something I hit when testing new things though to
my knowledge it's never been seen in released code.

Chrissie

**Comment by jnpkrn** [comment]
Basic fix for undesired segfault at libqb side (5th comment): #230

**Comment by jnpkrn** [comment]
To recap what's consequently happened:

- race-condition-between-independent-commits has been fixed: 0d6a698
- latent issue in a fail-case corner case has been fixed: #230
- corosync project notified about the latent segfault-handling-infloop:
  https://github.com/corosync/corosync/issues/159

That being said, this can likely be closed (please do if you can).

**Comment by chrissie-c** [comment]
Yes, closing this as the outstanding issues have been dealt with.
